lib/graphql/static_validation/literal_validator.rb

```diff
@@ -18,6 +18,22 @@ def validate(ast_value, type)
 
       private
 
+      def replace_nulls_in(ast_value)
+        case ast_value
+        when Array
+          ast_value.map { |v| replace_nulls_in(v) }
+        when Hash
+          new_v = {}
+          ast_value.each do |k, v|
+            new_v[k] = replace_nulls_in(v)
+          end
```

Comment:
@rmosolgo I might be missing something here, but doesn't this need to be this?

``` ruby
new_v = {}
ast_value.each do |k, v|
  new_v[k] = replace_nulls_in(v)
end
new_v
```

The `new_v` after the `each` is needed so that it gets returned rather than the what `ast_value.each { }` returns (which is the original `ast_value`, IIRC).

Or you could do one of the following:

``` ruby
ast_value.each_with_object({}) do |(k, v), new_v|
  new_v[k] = replace_nulls_in(v)
end

# or:
ast_value.to_h do |k, v|
  [k, replace_nulls_in(v)]
end
```

...either of which avoids the problem by using a method that automatically returns the new hash.

Also, it looks like the test you added isn't able to detect this since it doesn't actually use a hash in the data.  Might be worth adding a hash w/ a null value to the array in the test case.

Reply:
...actually, after writing that up I think there's an even simpler/better way to do this:

``` ruby
ast_value.transform_values { |v| replace_nulls_in(v) }
```

`transform_values` is like `map` for hashes, and seems ideal for this case.

Reply:
So strange... If I comment out my changes: 
```ruby
            ruby_value = ast_value # replace_nulls_in(ast_value)
            type.validate_input(ruby_value, @context)
```

And run my tests, then that test _does_ fail: 

```
$ be rake test TEST=spec/graphql/schema/build_from_definition_spec.rb

# Running tests with run options --seed 27810:

F

Finished tests in 0.018214s, 54.9028 tests/s, 219.6113 assertions/s.


Failure:
GraphQL::Schema::BuildFromDefinition::JSON type#test_0001_Sends normal ruby values to schema coercion [/Users/rmosolgo/code/graphql-ruby/spec/graphql/schema/build_from_definition_spec.rb:1709]
Minitest::Assertion: --- expected
+++ actual
@@ -1 +1,6 @@
-[nil, nil]
+[#<GraphQL::Language::Nodes::NullValue:0xXXXXXX
+  @col=33,
+  @filename=nil,
+  @line=2,
+  @name="null">,
+ nil]


1 tests, 4 assertions, 1 failures, 0 errors, 0 skips
```

But ... It _shouldn't've_ worked, right?!

Oh, I see your point about not using a Hash. Derp, just a minute...

Reply:
💁‍♂️ https://github.com/rmosolgo/graphql-ruby/pull/4324